### DISCUSSION TITLE
Combine blocking and non-blocking kafka retry unexpected behavior

### DISCUSSION THREAD

**Discussion by gemi2010** [spring-projects/spring-kafka#2669] [Q&A] [answered]
I'm combining blocking and non-blocking retry but i have an issue as follow:
When i add a certain exception to blocking with 

```
@Override
	protected void configureBlockingRetries(BlockingRetriesConfigurer blockingRetries) {
	    blockingRetries
	            .retryOn(BlockingException.class)
	            .backOff(new FixedBackOff(blockingRetryIntervalMillis, blockingMaxRetryAttempts));
	}
```

so it's blocking the consumer until the blockingMaxRetryAttempts are exhausted or it succeeded, but if the exception changed from BlockingException to NonBlockingException.class then i expect the flow to be switched to non-blocking retry as i configured 

```
@Bean
    public RetryTopicConfiguration retryableTopic(@Qualifier("retryableKafkaTemplate")KafkaTemplate<String, MtMessage> retryableKafkaTemplate) {	
        return RetryTopicConfigurationBuilder
                .newInstance()
                .fixedBackOff(nonblockingRetryIntervalMillis)
                .maxAttempts(nonblockingMaxRetryAttempts)
                .timeoutAfter(nonblockingRetryTimeout)
                .includeTopics(Arrays.asList("topic1"))
                .listenerFactory("retryKafkaListenerContainerFactory")
                .retryTopicSuffix(RETRY_TOPIC_SUFFIX)
                .dltSuffix(DLT_TOPIC_SUFFIX)
                .autoStartDltHandler(false)
                .useSingleTopicForSameIntervals()
                .doNotAutoCreateRetryTopics()
                .doNotRetryOnDltFailure()
                .retryOn(NonBlockingException.class)
                .create(retryableKafkaTemplate);
    }
```

but it didn't switch until the blocking retries were exhausted (let's say it retired 10 times blocking then in the 11th time the thrown exception is a non-blocking one so i expect to go to the non-blocking handler which it didn't happen), so how can i make it switch directly to non-blocking based on the configured exceptions on the `RetryTopicConfigurationBuilder. retryOn` ?

**Comment by garyrussell** [comment, marked as answer]
That was recently fixed (in 3.0.6 and 2.9.8) https://github.com/spring-projects/spring-kafka/pull/2645

**Reply by gemi2010** [reply]
Thanks, it works now as expected after i upgraded from 3.0.5 to 3.0.6
